### DISCUSSION TITLE
Breaking changes in 6.0 📣

### DISCUSSION THREAD

**Discussion by fbiville** [neo4j/neo4j-go-driver#456] [Announcements]
_Note: this is going to be updated on a regular basis._

_Note 2: release date for 6.0: Dec 2025_

If any item seems controversial, please join a discussion about this specific item (or create one if none exists).

## Removals

All of the items below must be marked as deprecated and documented with replacements.

### Protocol version

 - ~~Bolt 3 support will be dropped (all corresponding Neo4j server versions have long reached End of Life)~~ No longer valid. 6.0 will still negotiate Bolt 3.

### Structs

- `neo4j.Config` (replaced by `config.Config` since 5.8.0)

### Types

- `neo4j.ServerAddressResolver` (replaced by `config.ServerAddressResolver` since 5.8.0)
- `neo4j.LogLevel` (replaced by `log.Level` since 5.17.0)
- `neo4j.ERROR` (replaced by `log.ERROR` since 5.17.0)
- `neo4j.WARNING` (replaced by `log.WARNING` since 5.17.0)
- `neo4j.INFO` (replaced by `log.INFO` since 5.17.0)
- `neo4j.DEBUG` (replaced by `log.DEBUG` since 5.17.0)
- `log.Console` (replaced by `log.ToConsole` since 5.17.0)
- `log.Void` (replaced by `log.ToVoid` since 5.17.0)
- `neo4j.NotificationSeverity` (replaced by `notifications.NotificationSeverity` since 5.23.0)
- `neo4j.Warning` (replaced by `notifications.Warning` since 5.23.0)
- `neo4j.Information` (replaced by `notifications.Information` since 5.23.0)
- `neo4j.UnknownSeverity` (replaced by `notifications.UnknownSeverity` since 5.23.0)
- `neo4j.NotificationCategory` (replaced by `notifications.NotificationCategory` since 5.23.0)
- `neo4j.Hint` (replaced by `notifications.Hint` since 5.23.0)
- `neo4j.Unrecognized` (replaced by `notifications.Unrecognized` since 5.23.0)
- `neo4j.Unsupported` (replaced by `notifications.Unsupported` since 5.23.0)
- `neo4j.Performance` (replaced by `notifications.Performance` since 5.23.0)
- `neo4j.Deprecation` (replaced by `notifications.Deprecation` since 5.23.0)
- `neo4j.Generic` (replaced by `notifications.Generic` since 5.23.0)
- `neo4j.Security` (replaced by `notifications.Security` since 5.23.0)
- `neo4j.Topology` (replaced by `notifications.Topology` since 5.23.0)
- `neo4j.UnknownCategory` (replaced by `notifications.UnknownCategory` since 5.23.0)

### Struct fields

- `neo4j.Config#RootCAs` (this can be configured through `neo4j.Config#TlsConfig` since 5.0.0)

### Functions / Methods

- `neo4j.NewDriver` (replaced by `neo4j.NewDriverWithContext` since 5.0.0)
- `neo4j.SingleT` (replaced by `neo4j.SingleTWithContext` since 5.0.0)
- `neo4j.Single` (replaced by `neo4j.SingleWithContext` since 5.28.2)
- `neo4j.CollectT` (replaced by `neo4j.CollectTWithContext` since 5.0.0)
- `neo4j.Collect` (replaced by `neo4j.CollectWithContext` since 5.0.0)
- `neo4j.ConsoleLogger` (replaced by `log.ToConsole` since 5.17.0)
- `neo4j.ConsoleBoltLogger` (replaced by `log.BoltToConsole` since 5.17.0)

### Interfaces

- `neo4j.Driver` (replaced by `neo4j.DriverWithContext` since 5.0.0)
- `neo4j.Session` (replaced by `neo4j.SessionWithContext` since 5.0.0)
- `neo4j.Transaction` (replaced by `neo4j.ExplicitTransaction` since 5.0.0)
- `neo4j.TransactionWork` (replaced by `neo4j.ManagedTransactionWork` since 5.0.0)
- `neo4j.Result` (replaced by `neo4j.ResultWithContext` since 5.0.0)
- `neo4j.ServerAddress` (replaced by `config.ServerAddress` since 5.8.0)

## Updates

- Bump Go baseline version to latest generally available for 6.0.0
  - Likely to be Go v1.24

**Comment by sp71** [comment]
Also should mention Node & Relationship ID will be [replaced](https://github.com/neo4j/neo4j-go-driver/blob/v5.28.0/neo4j/dbtype/graph.go#L30) by ElementID

**Reply by StephenCathcart** [reply]
Good catch @sp71, thank you 👍

**Reply by StephenCathcart** [reply]
FYI, we are not in a position to remove `Id`, `StartId`, `EndId` yet for 6.0; these are still deprecated and will be removed in some future version of the driver. See https://github.com/neo4j/neo4j-go-driver/pull/683/files

**Comment by StephenCathcart** [comment]
**Update (v5.28.2)**: Added `neo4j.Single` deprecation to the list above. This was a missed deprecation from the original result helper cleanup - `neo4j.Single` is now deprecated in favour of `neo4j.SingleWithContext` to complete the context-aware API migration.

Users on v5 should migrate to `neo4j.SingleWithContext` to prepare for v6.

Thanks 👍

**Comment by StephenCathcart** [comment]
6.0 shipped and we're on 6.2.0 now, closing this out.

If you're still migrating: [MIGRATION_GUIDE.md](https://github.com/neo4j/neo4j-go-driver/blob/6.x/MIGRATION_GUIDE.md) and the [upgrade guide](https://neo4j.com/docs/go-manual/current/upgrade/).
